akvo/rsr/usecases/change_project_parent.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+
+# Akvo Reporting is covered by the GNU Affero General Public License.
+# See more details in the license.txt file located at the root folder of the Akvo RSR module.
+# For additional details on the GNU license please see < http://www.gnu.org/licenses/agpl.html >.
+
+from django.db.models import Q
+from akvo.rsr.models import (
+    Project, RelatedProject, Result, Indicator, IndicatorPeriod, IndicatorDimensionName, IndicatorDimensionValue, DefaultPeriod
+)
+
+
+def get_lineage(project):
+    project_id = project.id
+    family = [project_id]
+    while True:
+        parent_id = Project.objects.filter(
+            Q(
+                related_projects__related_project=project_id,
+                related_projects__relation=RelatedProject.PROJECT_RELATION_CHILD
+            ) | Q(
+                related_to_projects__project=project_id,
+                related_to_projects__relation=RelatedProject.PROJECT_RELATION_PARENT
+            )
+        ).values_list('id', flat=True).first()
+        if parent_id is None:
+            break
+        family.append(parent_id)
+        project_id = parent_id
+    return Project.objects.filter(id__in=family)
+
+
+def get_nearest_common_ancestor(first_lineage, second_lineage):
+    for it in first_lineage:
+        if it in second_lineage:
+            return it
+
+
+def make_tree_from_list(items, parent_attr):
+    tree = []
+    lookup = {}
+    ids = [it['id'] for it in items]
+    for item in items:
+        item_id = item['id']
+        if item_id not in lookup:
+            lookup[item_id] = {'children': []}
+        lookup[item_id]['item'] = item
+        node = lookup[item_id]
+        parent_id = item.get(parent_attr, None)
+        if not parent_id or parent_id not in ids:
+            tree.append(node)
+        else:
+            if parent_id not in lookup:
+                lookup[parent_id] = {'children': []}
+            lookup[parent_id]['children'].append(node)
+    return tree
+
+
+def get_leave(branch):
+    return get_leave(branch['children'][0]) if len(branch['children']) else branch['item']
+
+
+def make_target_parent_map(tree, project_attr, object_id, target_id):
+    target_parent_map = {}
+    for node in tree:
+        if len(node['children']) < 2:
+            continue
```

Comment:
Do we want to log this case, when running in verbose mode? It would point to potential data errors?